src/main/java/ui/components/pickers/LabelPickerUILogic.java

```diff
@@ -1,382 +1,24 @@
-                        } else {
-                            topLabels.remove(label);
-                        }
-                    });
-        }
-    }
-
-    private boolean isInTopLabels(String name) {
-        // used to prevent duplicates in topLabels
-        return topLabels.stream()
-                .filter(label -> label.getActualName().equals(name))
-                .findAny()
-                .isPresent();
-    }
-
-    private void addRemovePossibleLabel(String name) {
-        // Deletes previous selection
-        if (targetLabel.isPresent()) {
-            // if there's a previous possible selection, delete it
-            // targetLabel can be
-            topLabels.stream()
-                    .filter(label -> label.getActualName().equals(targetLabel.get()))
-                    .findFirst()
-                    .ifPresent(label -> {
-                        if (issue.getLabels().contains(targetLabel.get()) || resultList.get(targetLabel.get())) {
-                            // if it is an existing label toggle fade and strike through
-                            label.setIsHighlighted(false);
-                            label.setIsFaded(false);
-                            if (resultList.get(label.getActualName())) {
-                                label.setIsRemoved(false);
-                            } else {
-                                label.setIsRemoved(true);
-                            }
-                        } else {
-                            // if not then remove it
-                            topLabels.remove(label);
-                        }
-                    });
-            targetLabel = Optional.empty();
-        }
-
-        if (!name.isEmpty()) {
-            // Try to add current selection
-            if (isInTopLabels(name)) {
-                // if it exists in the top pane
-                topLabels.stream()
-                        .filter(label -> label.getActualName().equals(name))
-                        .findFirst()
-                        .ifPresent(label -> {
-                            label.setIsHighlighted(true);
-                            if (issue.getLabels().contains(name)) {
-                                // if it is an existing label toggle fade and strike through
-                                label.setIsFaded(resultList.get(name));
-                                label.setIsRemoved(resultList.get(name));
-                            } else {
-                                // else set fade and strike through
-                                // if space is pressed afterwards, label is removed from topLabels altogether
-                                label.setIsFaded(true);
-                                label.setIsRemoved(true);
-                            }
-                        });
+    public LabelPickerState determineState(LabelPickerState initialState, Set<String> repoLabels, String userInput) {
```

Comment:
I wonder if repoLabels should be part of the state object. After all, they form part of the state.